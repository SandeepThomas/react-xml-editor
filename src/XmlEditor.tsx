import * as React from 'react';

import Bubble from './Bubble';
import Element from './Element';
import Parser from './Parser';
import { BubbleType, Actions, BubbleOptions, DocSpec, Xml } from './types';

interface Props {
    docSpec: DocSpec;
    ref: React.RefObject<XmlEditor>;
    xml: string;
}

interface State {
    bubble: BubbleOptions;
    xml?: Xml;
}

export default class XmlEditor extends React.Component<Props, State> {
    public constructor(props: Props) {
        super(props);
        this.setXml = this.setXml.bind(this);
        this.showBubble = this.showBubble.bind(this);
        this.state = {
            bubble: {
                attribute: '',
                element: '',
                id: '',
                left: 0,
                show: false,
                top: 0,
                type: BubbleType.ASKER,
                value: '',
            },
        };
    }

    public componentDidMount() {
        const { xml } = this.props;
        const parser = new Parser();
        parser.parseString(xml).then((result) => {
            this.setState({
                xml: result,
            });
        });
    }

    public render(): React.ReactNode {
        return (
            <React.Fragment>
                <div className="xonomy nerd">
                    { this.getRootNode() }
                </div>
                { this.getBubble() }
            </React.Fragment>
        );
    }

    public getXml() {
        return this.state.xml;
    }

    private setXml(xml: Xml) {
        this.setState({
            xml,
        });
    }

    private showBubble(askOptions: Partial<BubbleOptions>): void {
        this.setState((prevState: State) => ({
            bubble: Object.assign(prevState.bubble, askOptions),
        }));
    }

    private getActions(): Actions {
        return {
            setXml: this.setXml,
            showBubble: this.showBubble,
        };
    }

    private getBubble(): React.ReactNode {
        const { docSpec } = this.props;
        const { bubble, xml} = this.state;
        if (xml) {
            return (
                <Bubble
                    actions={ this.getActions() }
                    attribute={ bubble.attribute }
                    docSpec={ docSpec }
                    element={ bubble.element }
                    id={ bubble.id }
                    left={ bubble.left }
                    show={ bubble.show }
                    top={ bubble.top }
                    type={ bubble.type }
                    value={ bubble.value }
                    xml={ xml }
                />
            );
        }
        return null;
    }

    private getRootNode(): React.ReactNode {
        const { xml } = this.state;
        if (xml) {
            const key = Object.keys(xml)[0];
            return (
                <Element
                    actions={ this.getActions() }
                    attributes={ xml[key].$ }
                    childElements={ xml[key].$$ }
                    collapsed={ xml[key]['#collapsed'] }
                    id={ `${key}` }
                    name={ key }
                    xml={ xml }
                />
            );
        }
        return null;
    }
}
