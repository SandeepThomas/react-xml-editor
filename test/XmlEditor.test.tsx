import {describe, expect, test} from '@jest/globals';
import * as React from 'react';
import renderer from 'react-test-renderer';

import XmlEditor from '../src/XmlEditor';

describe('XmlEditor component', () => {
    test('xml editor renders properly', () => {
        const ref: React.RefObject<XmlEditor> = React.createRef();
        const component = renderer.create(
            <XmlEditor
                docSpec={{}}
                ref={ref}
                xml="<xml />"
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('getXml() returns xml', () => {
        const ref: React.RefObject<XmlEditor> = React.createRef();
        renderer.create(
            <XmlEditor
                docSpec={{}}
                ref={ref}
                xml="<xml />"
            />
        );
        expect(ref.current.getXml()).toBe(undefined);
    });

    test('click should hide bubble', () => {
        const component = renderer.create(
            <XmlEditor
                docSpec={{}}
                ref={React.createRef()}
                xml="<xml />"
            />
        );
        component.toJSON().props.onClick();
        expect(component.getInstance().state.bubble.show).toBe(false);
    });
});
