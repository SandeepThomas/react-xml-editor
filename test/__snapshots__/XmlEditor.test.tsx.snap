// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XmlEditor component xml editor renders properly 1`] = `
<div
  className="xonomy nerd"
  onClick={[Function]}
/>
`;
